[15-FEAT] 약속 제안 생성

cpplovelove opened the pull request:
## 작업 목록
- [x] 약속 제안 생성 기능 구현

## 세부 내용
- 생성 중에 필수값 검증해서 없는 경우, 존재하지 않는 categoryId 인 경우 API 문서 및 코드에 추가했습니다!

## 논의 사항
- routing-contoller 처음 사용해보는데 따로 루트 지정해줄 필요 없이 바로 컨트롤러로 들어와서 라우터 파일이 필요할 지? 의문이 들었어요🥹 제가 잘못 생각한 걸 수도 있쥐만 이 부분 같이 얘기해보고 싶어요!

- 서비스 폴더에 util-service 파일 만들어뒀는데 여러 군데에서 반복적으로 사용할 것 같은 함수 여기 모아서 사용하는 거 혹시 어떠신가요?! 


## 연결된 이슈
- #15

cpplovelove committed:
feat: 약속 제안 생성

-약속 제안 생성 기능 구현

cpplovelove committed:
fix: 약속 제안 생성

- routing-contorller 적용

cpplovelove committed:
feat: 에러 처리 코드 구현

- 약속 제안 생성 중 발생할 수 있는 예외 처리
- util-service > paramValidation으로 필수값 검증

binimini commented on controllers/promising-controller.ts:
```diff
@@ -0,0 +1,21 @@
+
+
+@Controller()
+@UseBefore(bodyParser())
+export class PromisingController {
```
`app.use(bodyParser.json())` 도 명시되어있던데 중복으로 선언해야하는 부분인지 확인부탁드려요!
제가 구현할 때는 다음과 같이 설정했습니다!
```ts
import bodyParser from 'body-parser';

const app = express();

app.use(bodyParser.urlencoded());
app.use(bodyParser.json());

useExpressServer(app, {
  routePrefix: '/api',
  controllers: [UserController]
});
```

cpplovelove replied:
컨트롤러 쪽에서 바디파서 사용해주지 않으면 바디가 파싱 안되더라구요ㅠㅠ app.ts에 선언한 바디파서 사용부분 삭제했습니다!

binimini replied:
오잉 저는 저렇게 해서 POST 요청 받았는데....? 뭘까요 🤔

cpplovelove replied:
엇 받아지셨나요?! 제가 먼가 잘못 봤나봐요ㅠㅠ 그럼 한번 app.ts 에서 사용할 수 있도록 해보겠습니다!

binimini commented on controllers/promising-controller.ts:
```diff
@@ -0,0 +1,21 @@
+export class PromisingController {
+    @Post('/promisings')
+    async create(@Req() req: Request, @Res() res: Response) {
+        let resJson = {}
+        try {
+            resJson = await PromisingService.create(req, res)
+            return resJson
+        } catch (err: any) {
+            return err
+        }
```
resJson은 어떤 것을 반환해주는지에 대한 정보가 명확히 드러나지 않는 것 같아요.
promising을 생성해 받는 코드인 것을 나타내주는 게 어떨까요?
또 Response를 사용하는 김에 반환시의 상태와 전달 값을 좀 더 명확히 표현하는 게 좋을 것 같아요.

```js
promising = await PromisingService.create(req, res)
return res.status(200).send(promising);
```

cpplovelove replied:
훨씬 알아보기 좋은 것 같아요! 수정하겠습니다 ㅎㅎ

binimini commented on services/util-service.ts:
```diff
@@ -0,0 +1,38 @@
+const UtilService = {
+    convertTime(dateTime: Date) {
```
저는 개인적으로 UtilService라고하면 공통된 비즈니스 로직을 맡은 의존 객체를 늘리는 느낌이라서
가볍게 각각이 분리된 느낌으로 함수로 작성해서 export하는 편을 더 선호하긴 합니다 😺
`utils/index.ts`
```ts
export const convertTime:string = (dateTime: Date) => {};
//..
```

cpplovelove replied:
좋습니다! utils 폴더에 기능별로 파일 만들어서 분리했습니다! 파일이름 기능Util.ts로 통일하는거 괜찮으신가요?!

binimini replied:
파일 이름은 소문자로 해주세요~

binimini commented on routes/promising-router.ts:
```diff
@@ -0,0 +1,10 @@
+promisingRouter.post('/', PromisingController.create)
+
+export default promisingRouter;
+
```
라우팅이 필요없어져서 삭제하면 될 것 같아요~ 🙃

binimini commented on services/promising-service.ts:
```diff
@@ -0,0 +1,37 @@
+            const requireList = ['promisingName', 'ownerId', 'categoryId', 'minTime', 'maxTime'];
+            let promisingInfo = req.body;
+
+            const isValid = UtilService.paramValidation(promisingInfo, requireList)
```
사용자의 요청(request)와 응답(response)를 처리하는 부분은 Controller의 책임이지 않을까요?
우리가 Service를 나눠 구현하는 이유 중에 사용자의 직접적인 요청에 받는 영향을 최대한 줄이려는 부분도 있다고 생각해요.

Service에서 직접적으로 사용자의 요청을 그대로 사용하고 파싱하고 파라미터 검증하기보다는
그 부분은 Controller에게 맡기고 서비스의 로직이 보이게 작성해주시면 좋을 것 같아요!
```ts
async create(promisingName: string, ownerId: number, categoryId: number, minTime: Date, maxTime: Date) {
        
  const category = await CategoryKeyword.findOne({ where: { id: promisingInfo.categoryId } });
  const user = await User.findOne({ where: { id: promisingInfo.ownerId } });
  // .. 예외처리
  const promising = new PromisingModel(promisingInfo);
  return await promising.save();
}
```
그리고 Promising에 사용하는 사용자와 cateogry의 존재여부 검증은 
promising을 생성하고 저장하기 전에 처리해야하지 않을까요?
없는 사용자와 카테고리로 DB에 생성하는 코드를 일단 날리면 에러가 발생할 수 있을 것 같아요 😅

cpplovelove replied:
완전 좋은 지적 감사합니다! 예외 처리만 생각하고 순서를 전혀 생각 못했네요..해당 부분 수정해서 다시 커밋했습니다!

binimini commented on package.json:
```diff
@@ -34,9 +34,16 @@
     "typescript": "^4.6.4"
   },
   "dependencies": {
+    "@actions/core": "^1.8.2",
+    "@actions/github": "^5.0.3",
```
애네는 어디에 사용하는 의존성이에요? 😮

cpplovelove replied:
제가 삭제하는걸 깜빡한 의존성..^^입니다. 죄송해요ㅠㅠ 삭제해두겠습니다!

binimini commented on services/promising-service.ts:
```diff
@@ -0,0 +1,37 @@
+import { BadRequestError, NotFoundError, InternalServerError } from "routing-controllers";
+import User from "../models/user";
+
+const PromisingService = {
+    async create(req: Request, res: Response) {
```
Service로 생성할 것이라면 기존의 코드와 동일하게 class를 사용해서 생성하는 편이 나을 것 같아요. 
표현 방식이 매번 달라지면 혼란스러울 수 있으니까요. 🙃

```
class UserService {
  async func {}
}

export default new UserService();
```

binimini commented on services/util-service.ts:
```diff
@@ -0,0 +1,38 @@
+        jsonArray[target] = undefined;
+        return jsonArray;
+    },
+    paramValidation(paramList: any, requireList: any) {
+        for (const requirement of requireList) {
+            if (requirement in paramList == false) {
+                return false;
+            }
+        }
+        return true;
+    }
```
파라미터 검증을 이렇게 하는군요!
내부 메서드에서도 검증할 일이 있을 수도 있으니 구현해놓는 거 좋은 것 같아요 👍

cpplovelove replied:
일단 간단하게만 작성해뒀는데 더 좋은 방법 있다면 찾아서 수정해놓도록 하겠습니다! 조아요ㅎㅎ

binimini reviewed (changes requested):
고생하셨어요~~ 👍👍👍
말하신대로 라우팅 파일은 필요 없을 거 같아요. util도 있는 거 너무 좋아요! 
아무래도 `routing-controllers` 적용하면서 방식이 많이 달라져서 맞춰야할 부분이 많아졌네요 😹
열심히 같이 화이팅해봐욥 😘

cpplovelove committed:
fix: promising-controller 수정

- 중복 선언된 body-parser 삭제
- service파일 class로 통일
- 파라미터 검증과정 controller로 이동
- 필수 값 검증 코드 위치 변경

cpplovelove committed:
refactor: router 파일 삭제

- routing-controller 사용으로 필요없어진 router 파일 삭제

cpplovelove committed:
refactor: util 파일 구조 변경

- 하나의util-service에서 기능에 따라 util 폴더에 위치하도록 변경

binimini commented on services/promising-service.ts:
```diff
@@ -0,0 +1,30 @@
+import { NotFoundError, InternalServerError } from "routing-controllers";
+import User from "../models/user";
+
+class PromisingService {
+    async create(promisingInfo: any) {
+        try {
```
추가적으로 나중에 DTO 만들어서 타입 명시해주는 게 좋을 거 같아요
요렇게만 해놓으면 나중에 사용할 때 무슨 필드가 필요한지 찾기 어려울 수 있으니까!
중요한 리뷰는 거의 반영된 거 같아서 approve 해놓을게요~

binimini replied:
추가적으로 나중에 DTO 만들어서 타입 명시해주면 더 좋을 거 같아용
리뷰는 거의 다 반영된 거 같아서 일단 approve 해놓을게요!

binimini reviewed (approved):
👍👍

cpplovelove committed:
Merge branch 'dev' into feat-15-create-promising
